### PR TITLE
Implement fsp-pr, all SPL services, csrng. Rename LocationResolver functions.

### PR DESCRIPTION
**Pull request by SciresM**

### CODE REVIEW

File: nx/include/switch/services/csrng.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+ * @author SciresM
+ * @copyright libnx Authors
+ */
+#pragma once
+#include "../types.h"
+
+Result csrngInitialize(void);
+void csrngExit(void);
+
+Result csrngGetRandomBytes(void *out, size_t out_size);
```
Comment (fincs) [inline nx/include/switch/services/csrng.h]: Newline at eof

File: nx/include/switch/services/ldr.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        u32 aci0_sac_size;
+        u32 acid_fac_size;
+        u32 aci0_fah_size;
+        u8 ac_buffer[0x3E0];
+} LoaderProgramInfo;
+
+typedef struct {
+    u64 base_address;
+    u64 size;
+    unsigned char build_id[0x20];
```
Comment (fincs) [inline nx/include/switch/services/ldr.h]: Any reason why this is not just `u8`?

File: nx/include/switch/services/spl.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    SplConfigItem_DeviceId = 8,
+    SplConfigItem_BootReason = 9,
+    SplConfigItem_MemoryArrange = 10,
+    SplConfigItem_IsDebugMode = 11,
+    SplConfigItem_KernelMemoryConfiguration = 12,
+    SplConfigItem_IsChargerHiZModeEnabled = 13,
+    SplConfigItem_IsKiosk = 14,
+    SplConfigItem_NewHardwareType = 15,
+    SplConfigItem_NewKeyGeneration = 16,
+    SplConfigItem_Package2Hash = 17
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Need trailing comma

File: nx/include/switch/services/spl.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+} SplConfigItem;
+
+typedef enum {
+    RsaKeyVersion_Deprecated = 0,
+    RsaKeyVersion_Extended = 1
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Need trailing comma

File: nx/include/switch/services/spl.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+void splEsExit(void);
+
+Result splFsInitialize(void);
+void splFsExit(void);
+
+Result splManuInitialize(void);
+void splManuExit(void);
+
+Result splGetConfig(SplConfigItem config_item, u64 *out_config);
+Result splUserExpMod(void *input, void *modulus, void *exp, size_t exp_size, void *dst);
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Const correctness for `input`, `modulus` and `exp`?

File: nx/include/switch/services/spl.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+Result splSetConfig(SplConfigItem config_item, u64 value);
+Result splGetRandomBytes(void *out, size_t out_size);
+Result splIsDevelopment(u8 *out_is_development);
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Should probably be `bool`

File: nx/include/switch/services/spl.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+Result splSetSharedData(u32 value);
+Result splGetSharedData(u32 *out_value);
+
+Result splCryptoGenerateAesKek(void *wrapped_kek, u32 key_generation, u32 option, void *out_sealed_kek);
+Result splCryptoLoadAesKey(void *sealed_kek, void *wrapped_key, u32 keyslot);
+Result splCryptoGenerateAesKey(void *sealed_kek, void *wrapped_key, void *out_sealed_key);
+Result splCryptoDecryptAesKey(void *wrapped_key, u32 key_generation, u32 option, void *out_sealed_key);
+Result splCryptoCryptAesCtr(void *input, void *output, size_t size, void *ctr);
+Result splCryptoComputeCmac(void *input, size_t size, u32 keyslot, void *out_cmac);
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Please revise const correctness in these 6 functions

File: nx/include/switch/services/spl.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+Result splCryptoLockAesEngine(u32 *out_keyslot);
+Result splCryptoUnlockAesEngine(u32 keyslot);
+Result splCryptoGetSecurityEngineEvent(Handle *out_event);
+
+Result splRsaDecryptPrivateKey(void *sealed_kek, void *wrapped_key, void *wrapped_rsa_key, size_t wrapped_rsa_key_size, RsaKeyVersion version, void *dst, size_t dst_size);
```
Comment (fincs) [inline nx/include/switch/services/spl.h]: Etc... (until end of file)

File: nx/source/services/fspr.c [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            u64 result;
+        } *resp = r.Raw;
+
+        rc = resp->result;
+    }
+    
+    return rc;
+}
+
+Result fsprSetCurrentProcess() {
```
Comment (fincs) [inline nx/source/services/fspr.c]: (void)

File: nx/source/services/ldr.c [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+#include "kernel/ipc.h"
+#include "services/ldr.h"
+#include "services/fs.h"
+#include "services/sm.h"
+
+static Service g_shellSrv, g_dmntSrv, g_pmSrv;
+static u64 g_shellRefCnt, g_dmntRefCnt, g_pmRefCnt;
+
+/* Service init/exit helpers. */
+Result _ldrSrvInitialize(Service *srv, u64 *refcnt, const char *name) {
```
Comment (fincs) [inline nx/source/services/ldr.c]: Internal functions (prefixed with `_`) should be `static`.

File: nx/include/switch/services/ldr.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+Result ldrShellInitialize(void);
+void ldrShellExit(void);
+
+Result ldrDmntInitialize(void);
+void ldrDmntExit(void);
+
+Result ldrPmInitialize(void);
+void ldrPmExit(void);
+
+Result ldrShellAddTitleToLaunchQueue(u64 tid, void *args, size_t args_size);
```
Comment (fincs) [inline nx/include/switch/services/ldr.h]: Should be `const void *args` if I'm not mistaken

File: nx/include/switch/services/ldr.h [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+Result ldrShellClearLaunchQueue(void);
+
+Result ldrDmntAddTitleToLaunchQueue(u64 tid, void *args, size_t args_size);
```
Comment (fincs) [inline nx/include/switch/services/ldr.h]: Should be `const void *args` if I'm not mistaken

File: nx/include/switch/services/ldr.h (hunk 2 above)
Comment (SciresM) [reply nx/include/switch/services/ldr.h]: Nope, will change.

File: nx/source/services/spl.c [hunk 13]
```diff
@@ -0,0 +1,1134 @@
...
+#include "kernel/ipc.h"
+#include "kernel/detect.h"
+#include "services/sm.h"
+#include "services/spl.h"
+
+static Service g_splSrv, g_splCryptoSrv, g_splSslSrv, g_splEsSrv, g_splFsSrv, g_splManuSrv;
+static u64 g_splRefCnt, g_splCryptoRefCnt, g_splSslRefCnt, g_splEsRefCnt, g_splFsRefCnt, g_splManuRefCnt;
+
+/* Helper prototypes for accessing handles. */
+static Service *_splGetGeneralSrv(void);
```
Comment (yellows8) [inline nx/source/services/spl.c]: Service*

File: nx/source/services/fspr.c [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        return 0;
+
+    Result rc = smGetService(&g_fsprSrv, "fsp-pr");
+
+    if (R_SUCCEEDED(rc) && kernelAbove400()) {
+        rc = fsprSetCurrentProcess();
+    }
+
+    return rc;
+
```
Comment (yellows8) [inline nx/source/services/fspr.c]: whitespace

File: nx/source/services/fspr.c [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    atomicIncrement64(&g_fsprRefCnt);
+
+    if (serviceIsActive(&g_fsprSrv))
+        return 0;
+
+    Result rc = smGetService(&g_fsprSrv, "fsp-pr");
+
+    if (R_SUCCEEDED(rc) && kernelAbove400()) {
+        rc = fsprSetCurrentProcess();
+    }
```
Comment (yellows8) [inline nx/source/services/fspr.c]: Should call fsprExit() here on failure.

File: nx/source/services/fspr.c [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+            u64 result;
+        } *resp = r.Raw;
+
+        rc = resp->result;
+    }
+    
+    return rc;
+}
+
+
```
Comment (yellows8) [inline nx/source/services/fspr.c]: extra newline

**Comment by yellows8** [comment]
"Service *" still used elsewhere.

File: nx/source/services/fspr.c [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+    return rc;
+
+}
+
+void fsprExit(void) {
+    if (atomicDecrement64(&g_fsprRefCnt) == 0)
+        serviceClose(&g_fsprSrv);
+}
+
+Result fsprRegisterProgram(u64 pid, u64 titleID, u64 storageID, const void *fs_access_header, size_t fah_size, const void *fs_access_control, size_t fac_size) {
```
Comment (yellows8) [inline nx/source/services/fspr.c]: Use FsStorageId.

File: nx/include/switch.h [hunk 18]
```diff
@@ -55,6 +55,7 @@ extern "C" {
 #include "switch/services/pm.h"
 #include "switch/services/set.h"
 #include "switch/services/lr.h"
+#include "switch/services/ldr.h"
```
Comment (fincs) [inline nx/include/switch.h]: This is redundant, it was already done in 48e94921c59

**Pull request merged**